app/models/namespace.rb

```diff
@@ -111,6 +130,88 @@ def ransackable_attributes(_auth_object = nil)
     def ransackable_associations(_auth_object = nil)
       %w[]
     end
+
+    def subtract_from_metadata_summary_count(namespaces, metadata_summary, by_one)
+      return if metadata_summary.empty?
+
+      update_metadata_summary_counts(namespaces, metadata_summary, by_one: by_one, addition: false)
+    end
+
+    def add_to_metadata_summary_count(namespaces, metadata_summary, by_one)
+      return if metadata_summary.empty?
+
+      update_metadata_summary_counts(namespaces, metadata_summary, by_one: by_one, addition: true)
+    end
+
+    private
+
+    def update_metadata_summary_counts(namespaces, metadata_summary, by_one: false, addition: true) # rubocop:disable Metrics/AbcSize,Metrics/MethodLength
+      metadata_summary_update_sql = build_metadata_summary_update_sql(metadata_summary,
+                                                                      addition ? Arel::Nodes::Addition : Arel::Nodes::Subtraction,
+                                                                      by_one:)
+
+      entry_table = Arel::Table.new(
+        Arel::Nodes::NamedFunction.new(
+          'jsonb_each',
+          [Namespace.arel_table[:metadata_summary].concat(metadata_summary_update_sql)]
+        ),
+        as: 'entry'
+      )
+
+      updated_metadata_summary = entry_table.project(
+        Arel::Nodes::NamedFunction.new(
+          'coalesce',
+          [
+            Arel::Nodes::NamedFunction.new('jsonb_object_agg', [entry_table[:key], entry_table[:value]]),
+            Arel::Nodes::InfixOperation.new('::', Arel::Nodes::Quoted.new('{}'), Arel::Nodes::SqlLiteral.new('jsonb'))
+          ]
+        )
+      ).where(Arel::Nodes::InfixOperation.new('::', entry_table[:value], Arel::Nodes::SqlLiteral.new('integer')).gt(0))
+
+      Namespace.transaction do
+        locked_namespaces = namespaces.lock('FOR UPDATE')
+        locked_namespaces.update_all(metadata_summary: Arel::Nodes::Grouping.new(updated_metadata_summary)) # rubocop:disable Rails/SkipsModelValidations
+      end
+    end
+
+    def build_metadata_summary_update_sql(metadata_summary, operation = Arel::Nodes::Addition, by_one: false)
+      metadata_summary_update_sql = metadata_summary.map do |metadata_field, count|
+        single_field_metadata_summary_jsonb(metadata_field, by_one ? 1 : count, operation)
+      end
+
+      Arel::Nodes::Grouping.new(Arel.sql(metadata_summary_update_sql.map(&:to_sql).join(' || ')))
+    end
+
+    def single_field_metadata_summary_jsonb(metadata_field, count, operation) # rubocop:disable Metrics/MethodLength
```

Comment:
This function is especially large so having it another class with a more complete description on what it is doing is needed.